File: src/main/java/com/hubsante/model/edxl/EdxlMessage.java
/**
 * Copyright © 2023-2026 Agence du Numerique en Sante (ANS)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hubsante.model.edxl;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@JsonPropertyOrder({
        "distributionID",
        "senderID",
        "dateTimeSent",
        "dateTimeExpires",
        "distributionStatus",
        "distributionKind",
        "descriptor",
        "content"})
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class EdxlMessage extends EdxlEnvelope {
    private List<ContentObject> content;

    public EdxlMessage() {
        super();
    }

    public EdxlMessage(String distributionID, String senderID, OffsetDateTime dateTimeSent, OffsetDateTime dateTimeExpires,
                       DistributionStatus distributionStatus, DistributionKind distributionKind, Descriptor descriptor,
                       ContentMessage innerMessage) {
        super(distributionID, senderID, dateTimeSent, dateTimeExpires, distributionStatus, distributionKind, descriptor);
        this.setContentFrom(innerMessage);
    }

    public EdxlMessage content(List<ContentObject> content) {
        this.content = content;
        return this;
    }

    public EdxlMessage addContentObject(ContentObject contentObject) {
        if (this.content == null) {
            this.content = new ArrayList<>();
        }
        this.content.add(contentObject);
        return this;
    }

    @JacksonXmlElementWrapper(useWrapping = true, localName = "content")
    @JsonProperty(value = "content")
    @JacksonXmlProperty(localName = "contentObject")
    @JsonInclude(value = JsonInclude.Include.ALWAYS)
    public List<ContentObject> getContent() {
        return content;
    }

    @JacksonXmlElementWrapper(useWrapping = true, localName = "content")
    @JsonProperty(value = "content")
    @JacksonXmlProperty(localName = "contentObject")
    @JsonInclude(value = JsonInclude.Include.ALWAYS)
    public void setContent(List<ContentObject> content) {
        this.content = content;
    }

    public <T extends ContentMessage> void setContentFrom(T embeddedContent) {
        List<ContentObject> contentObjectList = new ArrayList<>();
        contentObjectList.add(new ContentObject(new ContentWrapper(new EmbeddedContent(embeddedContent))));
        this.setContent(contentObjectList);
    }

    public ContentMessage getFirstContentMessage() {
        return this.getContent().stream().findFirst().orElseThrow(ArrayIndexOutOfBoundsException::new)
                .getContentWrapper().getEmbeddedContent().getMessage();
    }

    public List<ContentMessage> getAllContentMessages() {
        return this.content.stream()
                .map(contentObject -> ((ContentMessage) contentObject.getContentWrapper().getEmbeddedContent().getMessage()))
                .collect(Collectors.toList());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        EdxlMessage that = (EdxlMessage) o;
        return Objects.equals(content, that.content);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), content);
    }

    @Override
    public String toString() {
        return "class EdxlMessage {\n" +
                        "    distributionID: " + toIndentedString(super.getDistributionID()) + "\n" +
                        "    senderId: " + toIndentedString(super.getSenderID()) + "\n" +
                        "    dateTimeSent: " + toIndentedString(super.getDateTimeSent()) + "\n" +
                        "    dateTimeExpires: " + toIndentedString(super.getDateTimeExpires()) + "\n" +
                        "    distributionStatus: " + toIndentedString(super.getDistributionStatus()) + "\n" +
                        "    distributionKind: " + toIndentedString(super.getDistributionKind()) + "\n" +
                        "    descriptor: " + toIndentedString(super.getDescriptor()) + "\n" +
                        "    content: " + toIndentedString(content) + "\n" +
                "}";
    }

    /**
     * Convert the given object to string with each line indented by 4 spaces
     * (except the first line).
     */
    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }
}

File: generator_php-symfony/package/src/Model/edxl/EdxlMessage.php
<?php

namespace App\Model\edxl;

use DateTimeInterface;
use JMS\Serializer\Annotation as JMS;
use Symfony\Component\Validator\Constraints as Assert;

class EdxlMessage
{
    /**
     * @Assert\NotBlank
     * @JMS\Type("string")
     * @JMS\SerializedName("distributionID")
     */
    private ?string $distributionID = null;

    /**
     * @Assert\NotBlank
     * @JMS\Type("string")
     * @JMS\SerializedName("senderID")
     */
    private ?string $senderID = null;

    /**
     * @Assert\NotNull
     * @Assert\Type("\DateTimeInterface")
     * @JMS\Type("DateTimeImmutable<'Y-m-d\TH:i:sP'>")
     * @JMS\SerializedName("dateTimeSent")
     */
    private ?DateTimeInterface $dateTimeSent = null;

    /**
     * @Assert\NotNull
     * @Assert\Type("\DateTimeInterface")
     * @JMS\Type("DateTimeImmutable<'Y-m-d\TH:i:sP'>")
     * @JMS\SerializedName("dateTimeExpires")
     */
    private ?DateTimeInterface $dateTimeExpires = null;

    /**
     * @Assert\NotBlank
     * @JMS\Type("string")
     * @JMS\SerializedName("distributionStatus")
     */
    private ?DistributionStatus $distributionStatus = null;

    /**
     * @Assert\NotBlank
     * @JMS\Type("string")
     * @JMS\SerializedName("distributionKind")
     */
    private ?DistributionKind $distributionKind = null;

    /**
     * @Assert\NotNull
     * @JMS\Type("App\Model\Edxl\Descriptor")
     */
    private $descriptor;

    /**
     * @Assert\Valid
     * @Assert\Count(min=1)
     *
     * @JMS\Type("array<App\Model\Edxl\ContentObject>")
     * @JMS\XmlList(entry="contentObject", inline=false)
     * @JMS\SerializedName("content")
     */
    private array $content = [];

    public function __construct(
        ?string $distributionID = null,
        ?string $senderID = null,
        ?DateTimeInterface $dateTimeSent = null,
        ?DateTimeInterface $dateTimeExpires = null,
        ?DistributionStatus $distributionStatus = null,
        ?DistributionKind $distributionKind = null,
        $descriptor = null
    ) {
        $this->distributionID = $distributionID;
        $this->senderID = $senderID;
        $this->dateTimeSent = $dateTimeSent;
        $this->dateTimeExpires = $dateTimeExpires;
        $this->distributionStatus = $distributionStatus;
        $this->distributionKind = $distributionKind;
        $this->descriptor = $descriptor;
    }

    public function getDistributionID(): ?string { return $this->distributionID; }
    public function setDistributionID(?string $id): self { $this->distributionID = $id; return $this; }

    public function getSenderID(): ?string { return $this->senderID; }
    public function setSenderID(?string $id): self { $this->senderID = $id; return $this; }

    public function getDateTimeSent(): ?DateTimeInterface { return $this->dateTimeSent; }
    public function setDateTimeSent(?DateTimeInterface $dt): self { $this->dateTimeSent = $dt; return $this; }

    public function getDateTimeExpires(): ?DateTimeInterface { return $this->dateTimeExpires; }
    public function setDateTimeExpires(?DateTimeInterface $dt): self { $this->dateTimeExpires = $dt; return $this; }

    public function getDistributionStatus(): ?DistributionStatus { return $this->distributionStatus; }
    public function setDistributionStatus(?DistributionStatus $s): self { $this->distributionStatus = $s; return $this; }

    public function getDistributionKind(): ?distributionKind { return $this->distributionKind; }
    public function setDistributionKind(?DistributionKind $k): self { $this->distributionKind = $k; return $this; }

    public function getDescriptor() { return $this->descriptor; }
    public function setDescriptor($d): self { $this->descriptor = $d; return $this; }

    // ------------------------------------------------------------
    // Content Management
    // ------------------------------------------------------------

    /** @return ContentObject[] */
    public function getContent(): array
    {
        return $this->content;
    }

    /** @param ContentObject[] $content */
    public function setContent(array $content): self
    {
        $this->content = $content;
        return $this;
    }

    public function addContentObject(ContentObject $obj): self
    {
        $this->content[] = $obj;
        return $this;
    }

    public function setContentFrom(ContentMessage $message): self
    {
        $this->content = [
            new ContentObject(
                new ContentWrapper(
                    new EmbeddedContent($message)
                )
            )
        ];
        return $this;
    }

    public function getFirstContentMessage(): ContentMessage
    {
        if (empty($this->content)) {
            throw new \RuntimeException("No content items");
        }

        return $this->content[0]
            ->getContentWrapper()
            ->getEmbeddedContent()
            ->getMessage();
    }

    /** @return ContentMessage[] */
    public function getAllContentMessages(): array
    {
        return array_map(
            fn(ContentObject $obj) =>
                $obj->getContentWrapper()->getEmbeddedContent()->getMessage(),
            $this->content
        );
    }
}
